pkg/pillar/cmd/domainmgr/domainmgr.go

```diff
@@ -1138,6 +1319,21 @@ func handleCreate(ctx *domainContext, key string, config *types.DomainConfig) {
 		VmConfig:           config.VmConfig,
 		Service:            config.Service,
 	}
+
+	ctx.resourcesUsedByVMs[config.UUIDandVersion.UUID] = ResourcesUsedByVM{CPUSet: make(map[int]bool)}
+	// Not to get races with the other threads, that try to assign the CPUs at the moment, we have to lock here
+	// Will release after the corresponding control files are created and the arguments are passed to containerd
+	ctx.vmResourcesLock.Lock()
+	defer ctx.vmResourcesLock.Unlock()
```

Comment:
This defer doesn't execute until this function returns, and I don't think you want to hold it that long.
Hiding the lock inside a object and package avoids that.

Also, I don't know whether it is a good idea to hold the lock across the cgroup updates (I have no idea how long the kernel can spend on that e.g., if some threads need to be stopped and rescheduled on a different CPU as part of processing the change).

What is key from a serialization perspective is to make sure that the manipulation of the maps is atomic, but once the assigned CPUs to the guest VM has been set and the shared map has been updated, both the booting of the guest with those pinned CPUs and the updating of the affinity for the others can be done separately. We just need to make sure that the updating of the affinity always gets the most recent shared set.

A golang pattern for that is to create a channel which is read in the main select loop, and you trigger that code by sending an empty message ("struct{}") conditionally on that channel. (We use that pattern to trigger sending protobuf messages in zedagent - look for "trigger" functions.)

Reply:
Done